Sources/TuistAutomation/Simulator/SimulatorController.swift

```diff
@@ -0,0 +1,107 @@
+import Foundation
+import RxSwift
+import TuistSupport
+
+protocol SimulatorControlling {
+    /// Returns the list of simulator devices that are available in the system.
+    func devices() -> Single<[SimulatorDevice]>
+
+    /// Returns the list of simulator runtimes that are available in the system.
+    func runtimes() -> Single<[SimulatorRuntime]>
+
+    /// Returns the list of simulator devices and runtimes.
+    func devicesAndRuntimes() -> Single<[SimulatorDeviceAndRuntime]>
+}
+
+enum SimulatorControllerError: FatalError {
+    case simctlError(String)
+
+    var type: ErrorType {
+        switch self {
+        case .simctlError: return .abort
+        }
+    }
+
+    var description: String {
+        switch self {
+        case let .simctlError(output): return output
+        }
+    }
+}
+
+final class SimulatorController: SimulatorControlling {
+    private let jsonDecoder: JSONDecoder = JSONDecoder()
+
+    func devices() -> Single<[SimulatorDevice]> {
+        System.shared.observable(["/usr/bin/xcrun", "simctl", "list", "devices", "--json"])
+            .mapToString()
+            .collectOutput()
+            .asSingle()
+            .flatMap { output in
+                guard output.standardError.isEmpty else {
+                    return .error(SimulatorControllerError.simctlError(output.standardError))
+                }
+                do {
+                    let data = output.standardOutput.data(using: .utf8)!
+                    let json = try JSONSerialization.jsonObject(with: data, options: [])
```

Comment:
is there a reason we can't map directly to a codable object?

Reply:
Yes, the structure is a bit odd:

```json
{
  "devices": {
    "runtime-id": [
      {
        // device
      }
    ]
  }
}
```
So I'm first extracting the device's dictionary, and from it the runtime IDs so that I can set them as attributes to the devices. I think this is a good example where Codable is not the best option.

Reply:
I'm not sure I fully understand, this should give you access to the identifier and devices - am I missing something?:

```swift
import Foundation

struct ListDevices: Codable {
    struct Device: Codable {
        var name: String
        var udid: String
    }
    var devices: [String: [Device]]
}

let list = try! JSONDecoder().decode(ListDevices.self, from: json.data(using: .utf8)!)
for (identifier, devices) in list.devices {
    identifier // com.apple.CoreSimulator.SimRuntime.tvOS-13-4
    devices[0].name // iPhone 11 Pro Max
}
```